File: src/main/java/edu/cornell/cs/apl/viaduct/errors/MalleableDowngradeError.java
package edu.cornell.cs.apl.viaduct.errors;

import edu.cornell.cs.apl.viaduct.imp.ast.DowngradeNode;
import java.io.PrintStream;

/**
 * Raised when downgrade nodes violate the non-malleable information flow control flow restriction.
 */
public class MalleableDowngradeError extends InformationFlowError {
  private final DowngradeNode node;

  public MalleableDowngradeError(DowngradeNode node) {
    this.node = node;
  }

  @Override
  protected String getCategory() {
    return "Malleable Downgrade";
  }

  @Override
  protected String getSource() {
    return node.getSourceLocation().getSourcePath();
  }

  @Override
  public void print(PrintStream output) {
    super.print(output);

    output.println("This downgrade node violates the non-malleability condition:");

    output.println();
    node.getSourceLocation().showInSource(output);
  }
}

File: src/main/kotlin/edu/cornell/cs/apl/viaduct/errors/MalleableDowngradeError.kt
package edu.cornell.cs.apl.viaduct.errors

import edu.cornell.cs.apl.prettyprinting.Document
import edu.cornell.cs.apl.viaduct.syntax.intermediate.DowngradeNode

/**
 * Thrown when a [DowngradeNode] node violate the non-malleable information flow control
 * restriction.
 */
class MalleableDowngradeError(private val node: DowngradeNode) : InformationFlowError() {
    override val category: String
        get() = "Malleable Downgrade"

    override val source: String
        get() = node.sourceLocation.sourcePath

    override val description: Document
        get() =
            // TODO: give more information to help diagnose the error?
            Document("This downgrade violates the non-malleability condition:")
                .withSource(node.sourceLocation)
}
